### DISCUSSION TITLE
Architecture going forward?  Entity model in this software or at consumer layer?

### DISCUSSION THREAD

**Discussion by spbrogan** [spbrogan/rvc2mqtt#9] [General]
A - Should this software do a simple dump everything into MQTT and then allow the MQTT consumer to build entities 

OR 

B - Should this software have a concept of entities (lights, water heater, furnace, temp sensor, etc) and then make that information available to a consumer. 
  - B-1 - Should this model use MQTT
  - B-2 - Should it do direct integration with HA or similar
  - B-3 - Do something else?  Produce it's own API?  

Additionally, how does the consumer create a "floor plan" without having to modify the main source of the project.  Floorplan would be used to translate RVC nodes into entites of relevance to the given trailer.  example DC_Load instance 1 == Bedroom light

**Comment by rperciaccante** [comment]
I am working on something similar, but the MQTT topics are already defined from the source of the event feed.  Unfortunately I have not yet dug into the CANbus side of the house, since it is proprietary and supposedly not compliant with the RV-C Standard.  Are you using a platform made by Lippert?

**Reply by spbrogan** [reply]
Keystone uses asa electronics in-command. 
https://www.keystonerv.com/innovations/in-command-smart-rv-control-system
https://in-command.net

The CAN Bus messages so far seem compliant with RV-C and have been easy to decode using the spec.  

Now that said I think what message type is used and how exactly the RV "devices" are addressed might be a bit crazy. Haven't done the work yet to map out all the nodes in the trailer and figure out exactly which message turns on/off/etc.  In a discussion with the author of the of https://github.com/linuxkidd/rvc-monitor-py he said that his trailer had a pretty odd mapping for some devices.  

In the next couple of days I am looking at integrating a "plugin" model.  Basically, allow generic RV-C decode and then pass the dictionary of key, values to a set of plugins to add/modify the keys/values.   

Then either write those out as topics and values or translate them into "objects" and then write out a clean/simple interface.  
Anyway, it is interesting to see how all this is constructed and what others have learned.

**Reply by rperciaccante** [reply]
This is similar the to the approach I have taken so far with the Node Red flows. In a future release I'll redo all the mqtt topics to keep things neat on both sides once I get to a working model with the intent to make things easy to extend.

**Reply by rperciaccante** [reply]
How do you intend to make the integration configurable? In my case, I can pull info from the Lippert Cloud Bridge that includes things like label, etc that helps bridge the gap between the consumer and the prosumer...

**Reply by spbrogan** [reply]
i am thinking right now that there are at least two levels of customizations. 

1. have a built-in plugin that will read a config yaml file authored by user that is invoking the app and will do simple translations of rvc data.  This would allow an easy way add a friendly name. 

2. User can supply a python plugin that gets the dictionary of key/value pairs on each message from Bus.  It could then translate it however needed to make the object (light, hvac, etc).  

Then those objects would be published using mqtt or otherwise.

**Comment by rlust** [comment]
Happy to help with testing! I have an Entegra Aspire 2022 and using the CP version from ERVIN on pi 3 with CAN, and running latest HomeAssistant on pi4. I am using the HUE integration on HA to control lights and awnings and Node-red in Home Assistant to read sensors with MQTT.

**Comment by spbrogan** [comment]
Have implemented the plugin system and authored 2 plugins.
* Light. - Using DC_LOAD_STATUS  - have not yet implemented MQTT control so it only reflects state in MQTT for now
* Temperature - simple temperature based on THERMOSTAT_AMBIENT_STATUS

Also made if fully configurable using a yaml file and you can even provide your own plugins so if your RV needs different behavior. 
 
Anyway looking for early feedback.  Docs are updated and should describe how to get started.

**Comment by spbrogan** [comment]
Added Tank Status as another plugin.

**Comment by rperciaccante** [comment]
Could you share your topic model?  I just completed onecontrol2mqtt and would like to see how you have the topics laid out, see if they are compatible, etc.

Thanks!

Bob

**Reply by spbrogan** [reply]
@rperciaccante  - Can you share your mqtt topic models?  For example, I am about to do the work for the hvac stuff and need to research what the topics should be.

**Reply by rperciaccante** [reply]
@spbrogan Well, here is what I am doing.

The base mqtt topic is configurable, so you can put it wherever you want.  In talking to some others, there was the question of how simple or complicated it should be, so I solved that problem - its up to you :)

That said, from there, I use the device name that I get from the controller, in this case it comes out as if it were a mqtt message, (but over http). 
```
topic: smarthome/items/idsmyrv_light_thing_0000000DA6BB1407_switch/state
payload: {"type":"OnOff","value":"OFF"}
```

I break that out like so:
```
topic: smarthome/items/idsmyrv_light_thing_0000000DA6BB1407_switch/state
baseTopic = smarthome/items
channelName = idsmyrv_light_thing_0000000DA6BB1407_switch
deviceName = light_0000000DA6BB1407
deviceID = 0000000DA6BB1407
entityType = switch
attribute = state
```

Since I get the feed from OpenHab, I have to convert to to something more generic
`new topic: (example) integration/onecontrol/device/switch_0000000DA7251E09/percent/`
there are 4 different kinds of messages: state, stateChange, deviceState, and command.  I split those up into topics off above:
```
light_0000000DA7251E09/switch/state
light_0000000DA7251E09/switch/stateChange
light_0000000DA7251E09/switch/deviceState
light_0000000DA7251E09/switch/command
```
(deviceState is the host device.  Think 3 relays on a board, if the board (device) goes down, the relays go offline)

Since the payloads of the different event types are different, I standardize them:
```
type: "onoff"    (decimal, string, onoff, percent, etc)
value: "off"
typeOrig: "OnOff"           (if during the standardization of the data it is changed, the old value is here)
valueOrig: "OFF"      (if during the standardization of the data it is changed, the old value is here)
details: ""                  (some values have additional details)
label: "Water Pump Relay Current Draw"  (this is the human readable name of the entity)
timeStamp: "Mon Feb 07, 2022 4:24:23pm"
```

Now that the inbound info is save, additional topic is created:
```
light_0000000DA7251E09/switch/attributes
```
While this is currently in progress, this will be used to hold information that is needed for other platforms, as well as info that is useful, but not necessarily worth a dedicated entity.  This is the info I will use for HA discovery and are retained.  When the flows are deployed, they pick up the attributes compare that info to the discovery host list.  If there is a device that is new, it adds it to the discovery topic to generate a new HA entity, if the device is not found, it removes it from discovery so the entity is deleted.

Probably a little more detailed than you intended, but i hope it helps.

**Comment by spbrogan** [comment]
I have some info here: 
https://github.com/spbrogan/rvc2mqtt/blob/main/docs/mqtt.md

I need to add more info for water heater and tank warmer.
